src/mixins/object_animation.mixin.ts

```diff
@@ -0,0 +1,184 @@
+// @ts-nocheck
+import { noop } from 'lodash';
+import { ObjectEvents } from '../EventTypeDefs';
+import { TDegree } from '../typedefs';
+import { animate } from '../util/animate';
+import { animateColor } from '../util/animate_color';
+import { StackedObject } from './object_ancestry.mixin';
+
+/**
+ * TODO remove transient
+ */
+type TAnimationOptions = Record<string, any>;
+
+export abstract class AnimatableObject<
+  EventSpec extends ObjectEvents = ObjectEvents
+> extends StackedObject<EventSpec> {
+  /**
+   * Animation duration (in ms) for fx* methods
+   * @type Number
+   * @default
+   */
+  FX_DURATION: number;
+
+  /**
+   * List of properties to consider for animating colors.
+   * @type String[]
+   */
+  colorProperties: string[];
+
+  abstract rotate(deg: TDegree): void;
+
+  /**
+   * Animates object's properties
+   * @param {String|Object} property Property to animate (if string) or properties to animate (if object)
+   * @param {Number|Object} value Value to animate property to (if string was given first) or options object
+   * @tutorial {@link http://fabricjs.com/fabric-intro-part-2#animation}
+   * @return {AnimationContext | AnimationContext[]} animation context (or an array if passed multiple properties)
+   *
+   * As object — multiple properties
+   *
+   * object.animate({ left: ..., top: ... });
+   * object.animate({ left: ..., top: ... }, { duration: ... });
+   *
+   * As string — one property
+   *
+   * object.animate('left', ...);
+   *
+   */
+  animate<T>(key: string, toValue: T, options?: TAnimationOptions): void;
+  animate<T>(animatable: Record<string, T>, options?: TAnimationOptions): void;
+  animate<T, S extends string | Record<string, T>>(
+    arg0: S,
+    arg1: S extends string ? T : TAnimationOptions,
+    arg2?: TAnimationOptions
+  ) {
+    const animatable = (
+      typeof arg0 === 'string' ? { [arg0]: arg1 } : arg0
+    ) as Record<string, T>;
+    const keys = Object.keys(animatable);
+    const options = (
+      typeof arg0 === 'string' ? arg2 : arg1
+    ) as TAnimationOptions;
+    keys.map((key, index) =>
+      this._animate(
+        key,
+        animatable[key],
+        index === keys.length - 1
+          ? options
+          : { ...options, onChange: undefined, onComplete: undefined }
+      )
+    );
+  }
+
+  /**
+   * @private
+   * @param {String} key Property to animate
+   * @param {String} to Value to animate to
+   * @param {Object} [options] Options object
+   */
+  _animate<T>(key: string, to: T, options: TAnimationOptions = {}) {
```

Comment:
This method is so ugly
Once #8297 is merged I will rework this mixin to use an array animation perhaps... with color somehow